File: routes/sourceLabels.js
const express = require('express');
const db = require('../db');
const { requireAdmin } = require('../middleware/auth');
const router = express.Router();

router.get('/', (req, res) => {
  try {
    const labels = db.prepare('SELECT * FROM source_labels ORDER BY name ASC').all();
    res.json(labels);
  } catch (e) {
    res.status(500).json({ error: e.message });
  }
});

router.post('/', requireAdmin, (req, res) => {
  try {
    const name = (req.body.name || '').trim();
    if (!name) return res.status(400).json({ error: 'Name is required' });

    const result = db.prepare('INSERT INTO source_labels (name) VALUES (?)').run(name);
    const label = db.prepare('SELECT * FROM source_labels WHERE id = ?').get(result.lastInsertRowid);
    res.status(201).json(label);
  } catch (e) {
    if (e.message.includes('UNIQUE')) return res.status(409).json({ error: 'Label already exists' });
    res.status(500).json({ error: e.message });
  }
});

router.delete('/:id', requireAdmin, (req, res) => {
  try {
    const existing = db.prepare('SELECT * FROM source_labels WHERE id = ?').get(req.params.id);
    if (!existing) return res.status(404).json({ error: 'Label not found' });
    db.prepare('DELETE FROM source_labels WHERE id = ?').run(req.params.id);
    res.json({ message: 'Label deleted successfully' });
  } catch (e) {
    res.status(500).json({ error: e.message });
  }
});

module.exports = router;

File: lib/routes/source_labels.dart
import 'dart:convert';
import 'package:shelf/shelf.dart';
import 'package:shelf_router/shelf_router.dart';
import '../database.dart';
import '../auth_middleware.dart';

void registerSourceLabelRoutes(Router router) {
  // Get all source labels (public)
  router.get('/api/source-labels', (Request request) async {
    try {
      final labels = queryAll('SELECT * FROM source_labels ORDER BY name ASC');
      return jsonResponse(labels);
    } catch (e) {
      return jsonResponse({'error': e.toString()}, statusCode: 500);
    }
  });

  // Create source label (admin only)
  router.post('/api/source-labels', (Request request) async {
    final authError = requireAdmin(request);
    if (authError != null) return authError;

    try {
      final body = jsonDecode(await request.readAsString()) as Map<String, dynamic>;
      final name = (body['name'] as String?)?.trim();
      if (name == null || name.isEmpty) {
        return jsonResponse({'error': 'Name is required'}, statusCode: 400);
      }

      execute('INSERT INTO source_labels (name) VALUES (?)', [name]);
      final label = queryOne('SELECT * FROM source_labels WHERE id = ?', [lastInsertRowId]);
      return jsonResponse(label, statusCode: 201);
    } catch (e) {
      if (e.toString().contains('UNIQUE')) {
        return jsonResponse({'error': 'Label already exists'}, statusCode: 409);
      }
      return jsonResponse({'error': e.toString()}, statusCode: 500);
    }
  });

  // Delete source label (admin only)
  router.delete('/api/source-labels/<id>', (Request request, String id) async {
    final authError = requireAdmin(request);
    if (authError != null) return authError;

    try {
      final labelId = int.parse(id);
      final existing = queryOne('SELECT * FROM source_labels WHERE id = ?', [labelId]);
      if (existing == null) {
        return jsonResponse({'error': 'Label not found'}, statusCode: 404);
      }
      execute('DELETE FROM source_labels WHERE id = ?', [labelId]);
      return jsonResponse({'message': 'Label deleted successfully'});
    } catch (e) {
      return jsonResponse({'error': e.toString()}, statusCode: 500);
    }
  });
}
